File: src/main/java/FormatGoMethodSignature.java
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.editor.actionSystem.EditorAction;
import com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler;

/**
 * @author Olivier Smedile
 * @author Vojtech Krasa
 */
public class FormatGoMethodSignature extends EditorAction {
    public FormatGoMethodSignature(boolean setupHandler) {
        super(null);
        if (setupHandler) {
            this.setupHandler(new EditorWriteActionHandler(true) {

                @Override
                public void executeWriteAction(Editor editor, Caret caret, DataContext dataContext) {
                    if (editor.isColumnMode()) {
                        return;
                    }

                    final SelectionModel selectionModel = editor.getSelectionModel();
                    boolean hasSelection = selectionModel.hasSelection();
                    if (!hasSelection) {
                        selectionModel.selectLineAtCaret();
                    }
                    final String selectedText = selectionModel.getSelectedText();

                    if (selectedText != null) {
                        final String newText = processSelection(selectedText);
                        applyChanges(editor, selectionModel, newText);
                    }
                }
            });

        }
    }

    private String processSelection(String selectedText) {
        return selectedText.replaceAll("\\((?!\\n)", "(\n\t")
                .replaceAll("(?<!\\n)\\)", ",\n)")
                .replaceAll(",(?!\\n)", ",\n\t");
    }

    public FormatGoMethodSignature() {
        this(true);
    }


    private void applyChanges(Editor editor, SelectionModel selectionModel, String newText) {
        editor.getDocument().replaceString(
                selectionModel.getSelectionStart(),
                selectionModel.getSelectionEnd(),
                newText);
    }
}

File: src/main/java/FormatGoMethodSignature.kt
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.editor.Caret
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.SelectionModel
import com.intellij.openapi.editor.actionSystem.EditorAction
import com.intellij.openapi.editor.actionSystem.EditorWriteActionHandler

/**
 * @author izumix03
 */
class FormatGoMethodSignature @JvmOverloads constructor(setupHandler: Boolean = true) : EditorAction(null) {
    init {
        if (setupHandler) {
            this.setupHandler(object : EditorWriteActionHandler(true) {
                override fun executeWriteAction(editor: Editor, caret: Caret?, dataContext: DataContext) {
                    if (editor.isColumnMode) {
                        return
                    }

                    val selectionModel = editor.selectionModel
                    if (!selectionModel.hasSelection()) {
                        selectionModel.selectLineAtCaret()
                    }
                    val selectedText = selectionModel.selectedText ?: return

                    val newText = processSelection(selectedText)
                    applyChanges(editor, selectionModel, newText)
                }
            })
        }
    }

    private fun processSelection(selectedText: String): String =
            selectedText.replace("\\((?!\\n)".toRegex(), "(\n\t")
                    .replace("(?<!\\n)\\)".toRegex(), ",\n)")
                    .replace(",(?!\\n)".toRegex(), ",\n\t")

    private fun applyChanges(editor: Editor, selectionModel: SelectionModel, newText: String) =
            editor.document.replaceString(
                    selectionModel.selectionStart,
                    selectionModel.selectionEnd,
                    newText)
}